


# Interesting yet Simple Functions in NumPy

### Exploring various Numpy functions for statistics, linear algebra, and more
> ### **Assignment 2 - Numpy Array Operations** 
>
> This assignment is part of the course ["Data Analysis with Python: Zero to Pandas"](http://zerotopandas.com). The objective of this assignment is to develop a solid understanding of Numpy array operations. 

NumPy is a Python library that provides many important functions which making working with arrays quicker and simpler. This notebook takes a look at a few functions in the library that could be useful solving complex problems in statistics, linear algebra, and more.

- np.transpose
- np.mean
- np.var
- np.where
- np.split


In [131]:
!pip install jovian --upgrade -q

In [132]:
import jovian

In [133]:
jovian.commit(project='numpy-array-operations')

<IPython.core.display.Javascript object>

[jovian] Updating notebook "abear9/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/abear9/numpy-array-operations


'https://jovian.ai/abear9/numpy-array-operations'

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [134]:
import numpy as np

In [135]:
# List of functions explained 
function1 = np.transpose #Produces transpose of a given array
function2 = np.mean #Find the mean of array or values along axis
function3 = np.var #Find the variance of array or values along axis
function4 = np.where #Returns values that match conditions given
function5 = np.split #Separates arrays into sub-arrays containing an equal amount of values

## Function 1 - np.transpose

np.transpose will produce a transposed array (i.e. flipping the array on its axes). 

In [138]:
# Example 1 - working
a1 = np.array([[0,1,2],[1,2,3]])
a1


array([[0, 1, 2],
       [1, 2, 3]])

In [139]:
a1.shape

(2, 3)

In [140]:
a2 = np.transpose(a1)
a2

array([[0, 1],
       [1, 2],
       [2, 3]])

In [141]:
a2.shape

(3, 2)

In the above example, we take a 2x3 array and transpose it to a 3x2 array. 

In [142]:
# Example 2 - working
a3 = np.arange(27).reshape(3,3,3)
a3

array([[[ 0,  1,  2],
        [ 3,  4,  5],
        [ 6,  7,  8]],

       [[ 9, 10, 11],
        [12, 13, 14],
        [15, 16, 17]],

       [[18, 19, 20],
        [21, 22, 23],
        [24, 25, 26]]])

In [143]:
a4 = np.transpose(a3,(1,0,2))
a4

array([[[ 0,  1,  2],
        [ 9, 10, 11],
        [18, 19, 20]],

       [[ 3,  4,  5],
        [12, 13, 14],
        [21, 22, 23]],

       [[ 6,  7,  8],
        [15, 16, 17],
        [24, 25, 26]]])

The above example makes use of the axes paramater on a 3D array. 

In [144]:
# Example 3 - breaking 
a5 = np.arange(6)
a5

array([0, 1, 2, 3, 4, 5])

In [145]:
a6 = np.transpose(a5,(1,2))
a6

ValueError: axes don't match array

If you are going to use the axes paramter, you must make sure the dimensions match the array you wish to transpose. Otherwise, you will get the above error. 

This function allows you to manipulate arrays for use in linear algebra or matrix mathematics. 

In [146]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "abear9/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/abear9/numpy-array-operations


'https://jovian.ai/abear9/numpy-array-operations'

## Function 2 - np.mean

np.mean allows you to find the mean of an array overall or along a certain axis. Some of the parameters for this function 
allow you to chose the type of number output, the axis along the mean is computed, or create a new array with the calculated mean values.

In [147]:
# Example 1 - working
x = np.array([[1,2,3],[4,5,6]])
np.mean(x)

3.5

Because no axis parameter is given, np.mean takes the mean of all of the values in the array. 

In [54]:
# Example 2 - working
np.mean(x,axis = 1)

array([2., 5.])

By setting the axis parameter to 1, np.mean takes the mean of the values of the items in the 1st dimension and creates an array
with shape (2,) containing the mean values.

In [148]:
# Example 3 - breaking (to illustrate when it breaks)
np.mean(x, axis = 3)

AxisError: axis 3 is out of bounds for array of dimension 2

Similar to the previous function, np.mean will break if the axis given is out of bounds for the array.

This is a useful function for finding the mean value of a whole set of values or individual lists in the array dataset.

In [56]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "abear9/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/abear9/numpy-array-operations


'https://jovian.ai/abear9/numpy-array-operations'

## Function 3 - np.var

This function returns the variance of array elements. By default, the function flattens the array, but an axis parameter may be given.

In [65]:
# Example 1 - working
x = np.arange(8)
x

array([0, 1, 2, 3, 4, 5, 6, 7])

In [66]:
np.var(x)

5.25

The above example takes the variance of all values in array x, which is 5.25. 

In [73]:
# Example 2 - working
x2 = np.random.rand(4,4)
x2

array([[0.43307025, 0.20569863, 0.29144896, 0.78427094],
       [0.15775739, 0.73718605, 0.12492873, 0.80832227],
       [0.72820595, 0.71865858, 0.60259015, 0.53447348],
       [0.44050272, 0.3741158 , 0.11376983, 0.26548779]])

In [77]:
np.var(x2, axis=1)

array([0.04875431, 0.10043725, 0.00658992, 0.01527439])

Setting the axis to 1 tells np.var to take the variance of each list in the array (the 1st axis). Each variance

In [80]:
# Example 3 - breaking 
np.var(x2, axis=3)

AxisError: axis 3 is out of bounds for array of dimension 2

Make sure that np.var's axis parameter is within the bounds for the array.

Variance is an important statistic when determining the spread between numbers in a data set. Whether it be measuring volatility in a market security, or determining product quality standards using a Six Sigma approach, NumPy's variance function provides a quick and simple way to calculate the statistic. 

In [81]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "abear9/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/abear9/numpy-array-operations


'https://jovian.ai/abear9/numpy-array-operations'

## Function 4 - np.where

np.where is a function that allows us to find values in an array that meet specific criteria. By default, the output array returns the index of such values, but the function allows us to pass parameter "x" which will replace values that satisfy the condition and parameter "y" which will replace values that do not. 

In [103]:
# Example 1 - working
x = np.random.randint(-100, high = 100, size = (20,))
x

array([ 65,  45,  16,  21, -20,  -5, -41, -83, -99,  75,  13,  61, -48,
        18, -93,  56,  -2, -93,  16,  84])

In [104]:
np.where(x >= 0)

(array([ 0,  1,  2,  3,  9, 10, 11, 13, 15, 18, 19]),)

The above example returns the index of each value greater than or equal to 0 in the array. It does not tell us the value, but rather only the index.

In [107]:
# Example 2 - working
np.where(x > 10, x, 'Less than 10')

array(['65', '45', '16', '21', 'Less than 10', 'Less than 10',
       'Less than 10', 'Less than 10', 'Less than 10', '75', '13', '61',
       'Less than 10', '18', 'Less than 10', '56', 'Less than 10',
       'Less than 10', '16', '84'], dtype='<U21')

For demonstrative purposes, this function checks whether each value is greater than 10, replaces the index with the value if it is greater than 10, or otherwise replaces the index with "Less than 10". 

In [108]:
# Example 3 - breaking (to illustrate when it breaks)
x2 = [1,2,3,4,5,6,7]
np.where(x2>4.5)

TypeError: '>' not supported between instances of 'list' and 'float'

The above example breaks because x2 is a normal Python list rather than a NumPy array. Because of this, Python cannot check the criteria for each value.

This function is useful for finding potential outliers, null values, or specific values in a dataset. In addition, np.where has the functionality to replace these values for an easier time analyzing the data.

In [109]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "abear9/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/abear9/numpy-array-operations


'https://jovian.ai/abear9/numpy-array-operations'

## Function 5 - np.split

This function will split an array into sub-arrays along a given axis. You may pass parameters for the amount of sub-arrays created or at certain indexes.

In [110]:
# Example 1 - working
x = np.arange(10)
np.split(x, 5)

[array([0, 1]), array([2, 3]), array([4, 5]), array([6, 7]), array([8, 9])]

This simple example takes an array with shape (10,) and splits it into 5 equal arrays. 

In [125]:
# Example 2 - working
x2 = np.arange(15)
np.split(x2, [2,6,10])

[array([0, 1]),
 array([2, 3, 4, 5]),
 array([6, 7, 8, 9]),
 array([10, 11, 12, 13, 14])]

In this example, np.split splits the array into sub-arrays for values up to index 2, values from indexes 2 up to 6, values from indexes 6 up to 10, and then the remaining values.

In [126]:
# Example 3 - breaking (to illustrate when it breaks)
x3 = np.arange(10)
np.split(x3, 4)

ValueError: array split does not result in an equal division

np.split will produce an error if splitting the array does not result in sub-arrays containing the same amount of values. Another function, i.e. array_split, can split the values correctly to fix this issue.

This function can be useful for breaking down larger arrays into smaller ones. By doing so, it could provide a better visual representation or an easier dataset to work with when conducting an analysis.

In [128]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "abear9/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/abear9/numpy-array-operations


'https://jovian.ai/abear9/numpy-array-operations'

## Conclusion

Overall, this notebook explores a few simple yet interesting functions that could be very useful for a variety of purposes. NumPy is quite powerful and contains an extensive library of useful tools. These tools help solve complex problems efficently and elegantly. 

## Reference Links
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* List of NumPy functions : https://numpy.org/doc/stable/reference/routines.html

In [149]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "abear9/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/abear9/numpy-array-operations


'https://jovian.ai/abear9/numpy-array-operations'